### PR TITLE
Stabilize copy_within

### LINKED ISSUE #54236: Tracking issue for copy_within
Rust PR: https://github.com/rust-lang/rust/pull/53652
Standalone implementation: https://github.com/oconnor663/copy_in_place

Summary: This is a safe wrapper around `ptr::copy`, for regions within a single slice. Previously, safe in-place copying was only available as a side effect of `Vec::drain`.

Open questions: @SimonSapin had a proposal for an alternative API using ranges. Here's an example PR along those lines: https://github.com/oconnor663/copy_in_place/pull/1

- [x] implement as a PR (https://github.com/rust-lang/rust/pull/53652)
- [x] land as unstable
- [ ] decide on a final API
- [ ] stabilize

### PR DESCRIPTION
**Pull request by kennytm**
Closes #54236.

### CODE REVIEW

**Comment by kennytm** [comment]
# Stabilization report

The `copy_within` function is a safe wrapper of `std::ptr::copy` within the slice's bound. This is an important building block in writing a fast and safe code involving moving multiple chunks of bytes in a `Vec<u8>`. 

```rust
let mut sl = [0, 1, 2, 3, 4, 5, 600, 700, 800, 900, 1000];
sl.copy_within(7..=9, 2);
assert_eq!(sl, [0, 1, 700, 800, 900, 5, 600, 700, 800, 900, 1000]);
```


The `copy_within` method was introduced in #53652, merged 8 months ago. Several APIs are explored in that PR, and the final form remains the same as today. This function is also available on crates.io as the [`copy_in_place` package](https://docs.rs/copy_in_place/0.2.0/copy_in_place/fn.copy_in_place.html), and its API also did not see any change after #53652 was merged. This shows an implicit consensus that the current function signature is good enough to stabilize as-is.

Test cases can be found in `src/libcore/tests/slice.rs`. It tests the normal behavior and error reporting when given invalid indices.

https://github.com/rust-lang/rust/blob/db4c783319d60ba2e4c93ac0b6a295f345a9c544/src/libcore/tests/slice.rs#L1494-L1538

There is a minor suggestion for [a non-panicking version](https://github.com/rust-lang/rust/issues/54236#issuecomment-447617175). However, all other slice methods like `rotate_left` and `split_at_mut` having panicking versions only, so this PR is not going to address that. If a non-panicking version is truly desired, we could introduce a `try_copy_within` method in the future.

**Comment by SimonSapin** [comment]
@rfcbot fcp merge

**Comment by SimonSapin** [comment]
Thanks for the stabilization report @kennytm !

**Comment by oconnor663** [comment]
> This shows an implicit consensus that the current function signature is good enough to stabilize as-is.

I'd caution that my crate has seen very little usage since it was published. I only use it in a couple places, and I only see one other calling crate on crates.io. So if someone did have other ideas about how the API should look, I wouldn't feel like we have a ton of usage evidence to argue either way. But on the other hand, it's a pretty simple API :)

I also notice that we might want to add tests for the overflow-panic cases that are possible with nonstandard ranges.

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -2146,14 +2146,13 @@ impl<T> [T] {
...
     ///
     /// bytes.copy_within(1..5, 8);
     ///
     /// assert_eq!(&bytes, b"Hello, Wello!");
     /// ```
-    #[unstable(feature = "copy_within", issue = "54236")]
+    #[stable(feature = "copy_within", since = "1.37.0")]
     pub fn copy_within<R: ops::RangeBounds<usize>>(&mut self, src: R, dest: usize)
     where
         T: Copy,
```
Comment (cramertj) [inline src/libcore/slice/mod.rs]: drive-by unrelated to stabilization: seems like we shouldn't be monomorphizing this whole function on `R`, given that we only need `.start_bound()` and `.end_bound()`, which we can call first and pass into a non-generic version.

Comment (SimonSapin) [reply src/libcore/slice/mod.rs]: Wouldn’t the method still be generic over `T`?

**Comment by kennytm** [comment]
Before the end of FCP I'd like to confirm one thing: would restricting the bound from `RangeBounds<usize>` to `SliceIndex<[T], Output = [T]>` be better? 

I don't think we're interested in supporting code like 

```rust
slice.copy_within((Bound::Excluded(1), Bound::Excluded(3)), 6);
``` 

The implementation could then also be shortened as:

```rust
pub fn copy_within<R: SliceIndex<[T], Output=[T]>>(&mut self, src: R, dest: usize) 
where
    T: Copy,
{
    let src = &self[src];
    let count = src.len();
    assert!(dest <= self.len() - count, "dest is out of bounds");
    unsafe {
        ptr::copy(
            src.as_ptr(),
            self.get_unchecked_mut(dest),
            count,
        );
    }
}
```

**Comment by SimonSapin** [comment]
This shortened code looks like it might have UB in case the source and destination ranges have overlap, since `&[T]` and `&mut [T]` exist at the same time. I’m not sure.

Regardless, being used for arguments like this is the point of `RangeBounds`. Even if we’re not interested in some of what it supports, there’s no need to actively exclude it.

**Comment by kennytm** [comment]
I've tried checking with miri but no errors were reported about overlapping ranges (unlike `&mut self[dest]` which was rejected).

The code should be changed to `self.as_mut_ptr().add(dest)` anyway, since both the existing code and the replacement above failed this in Miri:

```rust
    let mut x = [1, 2, 3];
    x.copy_within(1..1, 3);
```

File: src/libcore/slice/mod.rs (hunk 1 above)
Comment (nagisa) [reply src/libcore/slice/mod.rs]: LLVM will deduplicate identical functions anyway. Hard to see this coming at a compilation time cost and if it ever does in practice, it would be great to fix it in general rather than implementing hacks for specific functions in libcore.

File: src/libcore/tests/slice.rs [hunk 2]
```diff
@@ -1536,6 +1543,13 @@ fn test_copy_within_panics_src_inverted() {
     // 2 is greater than 1, so this range is invalid.
     bytes.copy_within(2..1, 0);
 }
+#[test]
+#[should_panic(expected = "attempted to index slice up to maximum usize")]
+fn test_copy_within_panics_src_out_of_bounds() {
+    let mut bytes = *b"Hello, World!";
+    // 2 is greater than 1, so this range is invalid.
```
Comment (oconnor663) [inline src/libcore/tests/slice.rs]:
```suggestion
    // an inclusive range ending at usize::max_value() would make src_end overflow
```

File: src/libcore/slice/mod.rs (hunk 1 above)
Comment (cramertj) [reply src/libcore/slice/mod.rs]: @nagisa the point is that they wouldn't be identical functions because they'd have different types for `R`. So we'd generate separate copies for `Range`, `RangeFrom`, `RangeTo`, `RangeInclusive`, etc...

**Comment by oconnor663** [comment]
I agree with @SimonSapin that putting `RangeBounds` in the public API is nice, since that's the stable trait that 3rd party types are allowed to implement. But I wish there was a helper function somewhere that converted from a `T: RangeBounds<usize>` to a `&[T]`, so that `copy_within` didn't need to do so many checks itself. Would it be worth factoring out such a function? Would it be worth _exposing_ such a function from the standard library in a separate PR?

**Comment by oconnor663** [comment]
This doesn't block stabilization, but something like [this refactoring](https://github.com/oconnor663/copy_in_place/commit/1c681b9ea697e2170810d0c729818f74c3fc6062) (example branch in the `copy_in_place` repo) could simplify things. In particular, it leaves us with just a single assert in front of the unsafe block. Here it is tweaked to be a method on slices:

```rust
pub fn copy_within<R: ops::RangeBounds<usize>>(&mut self, src: R, dest: usize)
where
    T: Copy,
{
    let (src_ptr, src_len) = {
        let src_slice = self.get_range(src); // <-- here's where a lot gets factored out
        (src_slice.as_ptr(), src_slice.len())
    };
    assert!(dest <= self.len() - src_len, "dest is out of bounds");
    unsafe {
        let dest_ptr = self.as_mut_ptr().add(dest);
        ptr::copy(src_ptr, dest_ptr, src_len);
    }
}
```

Here's [a godbolt output](https://godbolt.org/z/yA9mRW) suggesting that the optimizer is still happy.

File: src/libcore/slice/mod.rs (hunk 1 above)
Comment (nagisa) [reply src/libcore/slice/mod.rs]: @cramertj if the generated IR after inlining is identical, it will get merged. LLVM does not care about the type system.

Comment (cramertj) [reply src/libcore/slice/mod.rs]: My point is that I don't think the generated IR *would* be identical. `RangeFrom<T>` and `Range<T>` don't have the same representation, so converting them to start and end bounds would need to be split out into a separate function in order for the inner body of the functions to be unified.

**Comment by SimonSapin** [comment]
> I wish there was a helper function somewhere that converted from a `T: RangeBounds<usize>` to a `&[T]`

Is the `get` method of slices not that function?

----

Maybe forcing early conversion to raw pointers would could be enough to avoid UB?

```rust
let src: *const [T] = self.get(src);
let dest: *mut T = self.get_mut(dest);
```

**Comment by oconnor663** [comment]
> Is the `get` method of slices not that function?

I don't think there's a blanket impl of `SliceIndex` for `R: RangeBounds<usize>`, no. Unless I'm doing it wrong, I get "the type `[T]` cannot be indexed by `R`."

**Comment by SimonSapin** [comment]
I assume the nominated label was a request for FCP, which has happened.

@bors r+
